File: LeetCode 78 90-subset/duplicate_nums/subset_iteration.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        subsets = [[]]
        startIdx, endIdx = 0, 0
        
        for idx, num in enumerate(nums):
            startIdx = 0
            if idx > 0 and nums[idx] == nums[idx - 1]:
                startIdx = endIdx + 1
            endIdx = len(subsets) - 1
            
            for j in range(startIdx, endIdx+1):
                newSubset = subsets[j][::]
                newSubset.append(num)
                subsets.append(newSubset)
                
        return subsets

if __name__ == "__main__":
    solution = Solution()
    nums = [2, 1, 2]
    print("subsets with duplicate", solution.subsetsWithDup(nums))

# output: ('subsets with duplicate', [[], [1], [2], [1, 2], [2, 2], [1, 2, 2]])

File: LeetCode 78 90-subset/duplicate_nums/subset_iteration.go
package main

import (
	"fmt"
	"sort"
)

func subsetsWithDup(nums []int) [][]int {
	sort.Ints(nums)
	subsets := [][]int{{}}

	startIdx, endIdx := 0, 0
	for idx, num := range nums {
		startIdx = 0
		if idx > 0 && nums[idx] == nums[idx-1] {
			startIdx = endIdx + 1
		}
		endIdx = len(subsets) - 1

		for j := startIdx; j < endIdx+1; j++ {
			newSubset := append([]int{num}, subsets[j]...)
			subsets = append(subsets, newSubset)
		}
	}

	return subsets
}

func main() {
	nums := []int{2, 1, 2}
	fmt.Println("subsets with duplicate", subsetsWithDup(nums))
}

// output: subsets with duplicate [[] [1] [2] [2 1] [2 2] [2 2 1]]
